lib/spack/spack/solver/asp.py

```diff
@@ -1651,6 +1673,15 @@ def add_concrete_versions_from_specs(self, specs, origin):
                     DeclaredVersion(version=dep.version, idx=0, origin=origin)
                 )
                 self.possible_versions[dep.name].add(dep.version)
+                if (
+                    isinstance(dep.version, spack.version.GitVersion)
+                    and dep.version.user_supplied_reference
+                ):
+                    defined_version = spack.version.Version(dep.version.ref_version_str)
+                    self.declared_versions[dep.name].append(
+                        DeclaredVersion(version=defined_version, idx=0, origin=origin)
```

Comment:
@scheibelp: This should just ensure that all the CLI DeclaredVersions have a higher precedence than the others.  I think you needed the complex DeclaredVersion class above because you introduced a potentially duplicate `idx` here.

In general it's good for the weights in the concretizer (which the `idx` fields end up being used for) to be distinct, anyway, as it makes for deterministic answers.

Reply:
Update: setting the origin different (and restoring the definition of DeclaredVersion) causes unit test failures (although it avoids symmetries in the solver). I'm pretty sure setting the indices differently will run into the same failure. I will update on that here in the next hour.

Reply:
Note that this particular location is not the source of index-based equivalence since the versions are in fact not the same. That being said, I can change the index here if desired (it wouldn't be sufficient to address the index-based collisions that were occurring before, which is addressed by https://github.com/spack/spack/pull/36347/commits/c92938fe6afa6725388dd3e4c4234b793602675b)

Reply:
That is now done, so I will collapse this as resolved (the other comments sufficiently express the remaining concerns).